src/imports.rs

```diff
@@ -39,6 +35,24 @@ impl<'a> FmtVisitor<'a> {
             _ => ""
         };
 
+        if path_list.len() == 1 {
+            let vpi = path_list[0];
+            let rest = match vpi.node {
+                ast::PathListItem_::PathListIdent{ name, .. } => {
+                    token::get_ident(name).to_string()
+                }
+                ast::PathListItem_::PathListMod{ .. } => "self".to_string()
```

Comment:
You're right about `foo::{self}` paths, it's worth having a special case for that. I wouldn't panic! on `use {self}` I'd either leave it as it is is or output `use self` (probably worth checking that we don't introduce a compile error if there wasn't one before). My  reasoning is that if it is valid to parse it, then we shouldn't error out, at worst we should leave it unformatted.

No faff really :-)